File: src/main/java/uk/gov/justice/hmpps/offendersearch/health/HealthCheck.java
package uk.gov.justice.hmpps.offendersearch.health;


import lombok.AllArgsConstructor;
import org.springframework.boot.actuate.health.Health;
import org.springframework.boot.actuate.health.HealthIndicator;
import org.springframework.web.client.RestClientException;
import org.springframework.web.client.RestTemplate;

import static lombok.AccessLevel.PROTECTED;


@AllArgsConstructor(access = PROTECTED)
public abstract class HealthCheck implements HealthIndicator {
    private final RestTemplate restTemplate;

    @Override
    public Health health() {
        try {
            // TODO elasticsearch healthcheck
            return Health.up().build();
        } catch (final RestClientException e) {
            return Health.down(e).build();
        }
    }
}

File: src/main/kotlin/uk/gov/justice/hmpps/offendersearch/health/HealthCheck.kt
package uk.gov.justice.hmpps.offendersearch.health

import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.beans.factory.annotation.Value
import org.springframework.boot.actuate.health.Health
import org.springframework.boot.actuate.health.HealthIndicator
import org.springframework.stereotype.Component
import org.springframework.web.reactive.function.client.WebClient
import org.springframework.web.reactive.function.client.WebClientResponseException
import reactor.core.publisher.Mono
import java.time.Duration


abstract class HealthCheck(private val webClient: WebClient, private val timeout: Duration) : HealthIndicator {

  override fun health(): Health? {
    return webClient.get()
        .uri("/health/ping")
        .retrieve()
        .toEntity(String::class.java)
        .flatMap { Mono.just(Health.up().withDetail("HttpStatus", it?.statusCode).build()) }
        .onErrorResume(WebClientResponseException::class.java) { Mono.just(Health.down(it).withDetail("body", it.responseBodyAsString).withDetail("HttpStatus", it.statusCode).build()) }
        .onErrorResume(Exception::class.java) { Mono.just(Health.down(it).build()) }
        .block(timeout)
  }
}

@Component
class CommunityApiHealth
constructor(@Qualifier("communityApiHealthWebClient") webClient: WebClient, @Value("\${api.health-timeout:2s}") timeout: Duration) : HealthCheck(webClient, timeout)

@Component
class OAuthApiHealth
constructor(@Qualifier("oauthApiHealthWebClient") webClient: WebClient, @Value("\${api.health-timeout:2s}") timeout: Duration) : HealthCheck(webClient, timeout)

